packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
```

Comment:
pass `{generateNewIds: false}` or else the $modelIds of the clone will be different from the original

Reply:
Good point! Done in bb70a5b73446de0e62036699198461f25911a390.